### ISSUE TITLE
Monitor not working with Optifine

### ISSUE THREAD

**Issue by ferreusveritas** [dan200/ComputerCraft#534]
I'm pulling this into the proper area for action.  The original issue was posted on the Optifine issue tracker at https://github.com/sp614x/optifine/issues/1237   Please refer to it since it contains good information.

Basically the CC Monitor renders as an empty box under Optifine.   The cause is from having the Monitor 
TileEntity and TESR keep track of the last animation frame so it only renders once per frame.  I don't know why this is necessary.  Optifine renders in two passes.  The first is for volumes to create shadows and the second is for the actual surface rendering.  The second pass never occurs due to [this line](https://github.com/dan200/ComputerCraft/blame/914df8b0c7cc28451ea56a53808024511ff52642/src/main/java/dan200/computercraft/client/render/TileEntityMonitorRenderer.java#L56).

**Comment by SquidDev** [comment]
So the above commit _technically_ fixes this issue, but causes all sorts of other ones:

![](https://i.imgur.com/LuYbqwk.png)

It also seems to have a fun swing from side-to-side as you walk. Rendering is really not my forte, so if anybody has ideas I'd love to hear them.

**Comment by ferreusveritas** [comment]
Thanks for the quick response. Looks like the character quads are Z fighting with the background.  I'll see if I can reproduce it on my end. 
 I noticed that you didn't add any lighting code in that commit yet your screenshot shows a display that is clearly not black.  That strikes me as strange.  Did you do anything to fix that by chance?

**Comment by SquidDev** [comment]
I haven't touched the lighting code so was also rather surprised when it worked. Aside from a shaderpack I'm pretty sure I'm operating on default Optifine settings, so goodness knows what's going on. I don't know if it's specific to a particular set of shaders?

**Comment by ferreusveritas** [comment]
My "fix" was probably not the right way to go about it.  I'll test with multiple shaderpacks and see what I come up with.
Fyi, I was using the SEUS shader pack during testing since I believe it is the most popular.

**Comment by SquidDev** [comment]
Ahhh, I was using Sildurs Vibrant. My laptop has an Intel graphics card, and so has a burning hatred of SEUS.

**Comment by Lignum** [comment]
From what i can tell, ComputerCraft renders the character quads and the background quads directly on top of each other, and tries to prevent z fighting with:

```Java
            GlStateManager.depthMask( false );
```

No idea why, but this hack may not work with shaders for whatever reason. There shouldn't be any harm in getting rid of that, and simply moving the character quads a tiny bit to the front.

**Comment by ferreusveritas** [comment]
@SquidDev Sildurs Vibrant Lite I presume? Since your hash brown of a video card can't really handle anything more?

**Comment by SquidDev** [comment]
>  Looks like the character quads are Z fighting with the background. I'll see if I can reproduce it on my end.

>  From what i can tell, ComputerCraft renders the character quads and the background quads directly on top of each other, and tries to prevent z fighting with: `GlStateManager.depthMask( false );`

I had a bit more of a play and it doesn't appear that this is an issue with character z fighting: it occurs even with a flat background. Removing the depth mask setter (and shifting characters) has no effect on the end result.

![](https://i.imgur.com/Npxjj8e.png)

It's also worth noting that the player's shadow is substantially darker on monitors than any other block (you can see the change on the transition from border to monitor at the bottom).

> Sildurs Vibrant Lite I presume? Since your hash brown of a video card can't really handle anything more?

Yep.

**Comment by ferreusveritas** [comment]
Using MC 1.12.2 Sildurs Vibrant Lite on NVidia GTX 970 LinuxMint:

For these results I simply commented out the return on line 56.

The following image is with 
`OpenGlHelper.setLightmapTextureCoords(OpenGlHelper.lightmapTexUnit, 0, 0xFF);`
added before 
```
// Draw the contents
GlStateManager.depthMask( false );
```

![2018-03-29_09 20 22](https://user-images.githubusercontent.com/13294361/38091106-87b5637e-3332-11e8-9698-68271549a3c9.png)


This is without the setLightmapTextureCoords(Presumably how yours is setup)

![2018-03-29_09 20 30](https://user-images.githubusercontent.com/13294361/38091113-8a7ee21a-3332-11e8-81cc-e378aa5ddc03.png)

**Comment by SquidDev** [comment]
OK, if the artefacts are just due to my graphics card being awful then that's fine. Thanks for looking into it!

**Comment by ferreusveritas** [comment]
I was going to ask about that.  Are the monitor displays supposed to glow?  The setLightmapTextureCoords code I suggested adding does make the monitor glow regardless of daylight with shaders.  Even when underground in the dark.  I turned shaders OFF entirely and then disabled the setLightmapTextureCoords I added and it still glows underground.  So I figured that's just the way it's supposed to work since CRT/LCD/LED monitors are emissive devices IRL.  I'm guessing that at some point it was an added feature to "glow".
I'm kinda new to 1.12.2 and I didn't remember them ever doing that so I went back to MC 1.7.10 CC and the monitors don't glow in the dark.   So.. my mind.. it is boggled.

**Comment by SquidDev** [comment]
> I'm kinda new to 1.12.2 and I didn't remember them ever doing that so I went back to MC 1.7.10 CC and the monitors don't glow in the dark. So.. my mind.. it is boggled

Yeah, it was only added in 1.10.2. One of the first changes after being open sourced IIRC. Glad it's working though - I'll add that to the PR. I also need to check that rendering is correct when some monitors are outside the view frustrum - I'm not entirely sure my previous commit gets things entirety correct.

**Comment by ferreusveritas** [comment]
Setting the second parameter to 255 can correct the day/night thing but with some shader packs the monitor turns more amber(or whatever torchy color the pack provides) at night because of the color map applied during that time of day.

`OpenGlHelper.setLightmapTextureCoords(OpenGlHelper.lightmapTexUnit, 0xFF, 0xFF);`

Using this function was the only way I could get the color mapping to work right.  There's probably a better call to use that's not dependent on the time of day but I haven't found it yet.

**Comment by ferreusveritas** [comment]
> Yay, concurrent edits are confusing. I'm rather curious as to where the lightmap numbers you gave came from? I've had a prod through MC's code and it's rather like black magic.

The 255 values are just the maximum TorchLight(Block) and Daylight(Sky).  The values are used in a big lookup table in the form of a texture and can be changed via resourcepacks/shaders.
![lightmap](https://user-images.githubusercontent.com/13294361/38099160-e8154d6e-3347-11e8-9111-3d30c7439518.png)

**Comment by SquidDev** [comment]
Thanks for the explanation! That's pretty much what I'd assumed, but wasn't entirely sure of the layout - now it makes much more sense :). 

The issues linked on the Optifine tracker are definitely similar (sp614x/optifine#631 especially), and don't offer much hope. That being said, they do seem to largely refer to Forge's rendering pipeline, which isn't utilised here - `disableLightmap` is vanilla code. 

Anyway, that's rather outside the scope of this issue - I'll add the lightmap call to the PR :). Thank you so much for all your diagnostics help!

**Comment by ferreusveritas** [comment]
Thank you for carrying the CC torch as far as you have!  I'm very grateful for your efforts.  So is Lignumn's rewrite the most up to date version?

**Comment by Lignum** [comment]
> Edit: Using Lignumn's rewrite branch does prevent the black artifacts but still results in weird vertical warps. There's a download here if people want to experiment.

During the rewrite, I tried to minimise the amount of OpenGL state being set, as that usually leads to bugs. I found that a lot of the state changes ComputerCraft made were redundant (including `depthMask`). Since the rewrite isn't ready yet, I suggest we just extract the GL state code from the rewrite, and apply it here (might want to exclude `depthMask` from that), it should work the same, display lists or not. That still won't fix it, but if, as you say, it's less buggy, then that's a start.

> So is Lignumn's rewrite the most up to date version?

Not really, it's far from ready. It's really a rewrite of the entire terminal rendering code, creating a shared code base for all terminal-like displays, instead of copy & pasting it in everywhere, like it's currently done. Currently, computer terminals don't render correctly (should be easy to fix, I'm just lazy), and they lag a bit (this is due to some weirdness with detecting terminal changes, that's more difficult), so I don't recommend using it, it's far from stable.

**Comment by SquidDev** [comment]
I had a bit of a fiddle with some other lighting options, trying to see what other mods did, but to no avail. It's not optimal, but seems like the lightmap solution will have to do.

> So is Lignumn's rewrite the most up to date version?

Blughr, it's messy: I've got a branch (#507) which rewrites the monitors on the server and bundles some additional monitor fixes in order to avoid PR-dependency hell. Lignum's rewrites the client side handling of terminals (and so monitors).

> That still won't fix it, but if, as you say, it's less buggy, then that's a start.

I think this is just my graphics card being one Yorkshire pudding short of a roast. It only appears to occur on monitors facing one direction and there's all sorts of similar artefacts on normal blocks. Adding the lightmap change appears to fix it, so I don't think it's worth worrying about.

**Comment by Injen** [comment]
hi to all.
I was looking for a solution just like you in MC1.7.10
Im using SEUS v.11.

Here is:
Options -> Graphic Settings -> Shaders -> Shaders Settings -> Lighting and Shadow Options -> Global Illumination -> set OFF.
Done.

![image](https://user-images.githubusercontent.com/12650292/179797660-5ef8a5e2-101a-4a2f-9e50-c1d37f5e6fa0.png)
